pkg/endpoint/endpoint_status.go

```diff
@@ -0,0 +1,234 @@
+	"github.com/cilium/cilium/pkg/identity"
+	identitycache "github.com/cilium/cilium/pkg/identity/cache"
+	cilium_v2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+	"github.com/cilium/cilium/pkg/labels"
+	"github.com/cilium/cilium/pkg/policy/trafficdirection"
+)
+
+func getEndpointStatusControllers(status *models.EndpointStatus) (controllers cilium_v2.ControllerList) {
+	for _, c := range status.Controllers {
+		if c.Status == nil {
+			continue
+		}
+
+		if c.Status.ConsecutiveFailureCount > 0 {
+			s := cilium_v2.ControllerStatus{
+				Configuration: c.Configuration,
+				Name:          c.Name,
+				UUID:          string(c.UUID),
+				Status: cilium_v2.ControllerStatusStatus{
+					ConsecutiveFailureCount: c.Status.ConsecutiveFailureCount,
+					FailureCount:            c.Status.FailureCount,
+					LastFailureMsg:          c.Status.LastFailureMsg,
+					LastFailureTimestamp:    c.Status.LastFailureTimestamp.String(),
+					LastSuccessTimestamp:    c.Status.LastSuccessTimestamp.String(),
+					SuccessCount:            c.Status.SuccessCount,
+				},
+			}
+			if controllers == nil {
+				controllers = cilium_v2.ControllerList{s}
+			} else {
+				controllers = append(controllers, s)
+			}
+		}
+	}
+
+	if controllers != nil {
+		controllers.Sort()
+	}
+
+	return
+}
+
+func (e *Endpoint) getEndpointStatusLog() (log []*models.EndpointStatusChange) {
+	added := 0
+
+	if s := e.Status; s != nil {
+		s.indexMU.RLock()
+		defer s.indexMU.RUnlock()
+
+		for i := s.lastIndex(); ; i-- {
+			if i < 0 {
+				i = maxLogs - 1
+			}
+			if i < len(s.Log) && s.Log[i] != nil {
+				l := &models.EndpointStatusChange{
+					Timestamp: s.Log[i].Timestamp.Format(time.RFC3339),
+					Code:      s.Log[i].Status.Code.String(),
+					Message:   s.Log[i].Status.Msg,
+					State:     models.EndpointState(s.Log[i].Status.State),
+				}
+
+				if strings.ToLower(l.Code) != models.EndpointStatusChangeCodeOk {
+					if log == nil {
+						log = []*models.EndpointStatusChange{l}
+					} else {
+						log = append(log, l)
+					}
+
+					// Limit the number of endpoint log
+					// entries to keep the size of the
+					// EndpointStatus low.
+					added++
+					if added >= cilium_v2.EndpointStatusLogEntries {
+						break
+					}
+				}
+			}
+			if i == s.Index {
+				break
+			}
+		}
+	}
+	return
+}
+
+func getEndpointIdentity(status *models.EndpointStatus) (identity *cilium_v2.EndpointIdentity) {
+	if status.Identity != nil {
+		identity = &cilium_v2.EndpointIdentity{
+			ID: status.Identity.ID,
+		}
+
+		identity.Labels = make([]string, len(status.Identity.Labels))
+		copy(identity.Labels, status.Identity.Labels)
+		sort.Strings(identity.Labels)
+	}
+	return
+}
+
+func getEndpointNetworking(status *models.EndpointStatus) (networking *cilium_v2.EndpointNetworking) {
+	if status.Networking != nil {
+		networking = &cilium_v2.EndpointNetworking{
+			Addressing: make(cilium_v2.AddressPairList, len(status.Networking.Addressing)),
+		}
+		i := 0
```

Comment:
i is never incremented , so you will be overwriting prior values in `status.Networking.Addressing`. I can't imagine us ever having more than one IPv4 or IPv6 address, though.